server/tests/api_v4/authenticated/test_vlob_update.py

```diff
@@ -249,4 +250,218 @@ async def test_authenticated_vlob_update_bad_vlob_version(
+    coolorg: CoolorgRpcClients, backend: Backend
+) -> None:
+    raise Exception("Not implemented")
+
+
+async def test_authenticated_vlob_update_timestamp_out_of_ballpark(
+    coolorg: CoolorgRpcClients, backend: Backend
+) -> None:
+    initial_dump = await backend.vlob.test_dump_vlobs(organization_id=coolorg.organization_id)
+    t0 = DateTime.now().subtract(seconds=3600)
+    rep = await coolorg.alice.vlob_create(
+        realm_id=coolorg.wksp1_id,
+        vlob_id=VlobID.new(),
+        key_index=1,
+        timestamp=t0,
+        blob=b"<block content>",
+        sequester_blob=None,
+    )
+    assert isinstance(rep, authenticated_cmds.v4.vlob_create.RepTimestampOutOfBallpark)
+    assert rep.ballpark_client_early_offset == 300.0
+    assert rep.ballpark_client_late_offset == 320.0
+    assert rep.client_timestamp == t0
+
+    # Ensure no changes were made
+    dump = await backend.vlob.test_dump_vlobs(organization_id=coolorg.organization_id)
+    assert dump == initial_dump
+
+
+@pytest.mark.parametrize(
+    "timestamp_offset",
+    (pytest.param(0, id="same_timestamp"), pytest.param(1, id="previous_timestamp")),
+)
+async def test_authenticated_vlob_create_require_greater_timestamp(
+    coolorg: CoolorgRpcClients,
+    backend: Backend,
+    timestamp_offset: int,
+) -> None:
+    dt1 = DateTime.now()
+    dt2 = dt1.add(seconds=10)
+    bad_timestamp = dt2.subtract(seconds=timestamp_offset)
+
+    # 1) Create a first vlob
+
+    author = coolorg.alice.device_id
+    realm_id = coolorg.wksp1_id
+    organization_id = coolorg.organization_id
+    vlob_id = VlobID.new()
+    outcome = await backend.vlob.create(
+        now=dt1,
+        organization_id=organization_id,
+        author=author,
+        realm_id=coolorg.wksp1_id,
+        vlob_id=vlob_id,
+        key_index=1,
+        blob=b"<initial block content>",
+        timestamp=dt1,
+        sequester_blob=None,
+    )
+    assert outcome is None, outcome
+
+    initial_dump = await backend.vlob.test_dump_vlobs(organization_id=organization_id)
+
+    outcome = await backend.realm.rotate_key(
+        now=dt2,
+        organization_id=organization_id,
+        author=author,
+        author_verify_key=coolorg.alice.signing_key.verify_key,
+        keys_bundle=b"",
+        per_participant_keys_bundle_access={
+            coolorg.alice.user_id: b"",
+            coolorg.bob.user_id: b"",
+        },
+        realm_key_rotation_certificate=RealmKeyRotationCertificate(
+            author=author,
+            timestamp=dt2,
+            hash_algorithm=HashAlgorithm.SHA256,
+            encryption_algorithm=SecretKeyAlgorithm.XSALSA20_POLY1305,
+            key_index=2,
+            realm_id=realm_id,
+            key_canary=b"",
+        ).dump_and_sign(coolorg.alice.signing_key),
+    )
+    assert isinstance(outcome, RealmKeyRotationCertificate)
+
+    # 2) Create second vlob with same or previous timestamp
```

Comment:
the comment should be about updating the vlob

Reply:
Done